README
                       +---------------+
                       | bbbutton v0.8 |
                       +---------------+
	
            HOMEPAGE: www.angelfire.com/theforce/button/

	   
1. What is bbbutton

    This is a dock application for BlackBox (not only), that allows you
    to create single application launch buttons with the image and
    application launch of your preference.
    
2. Install/Uninstall/Upgrade

    tar -zxvf bbbutton_0.8.tar.gz
    cd bbbutton_0.8
    make <systype>        (systypes recognised: linux, freebsd)
    make install ( or uninstall )
    
    NOTE: To upgrade to version 0.8 replace your existing "bbbutton"
          binary, but remember to use the new .bbbuttonrc format
          as outlined in the next section.

3. OK, now what !!
  
   Load multiple buttons when your X windows start using the 
   configuration file .bbbuttonrc which is located 
   in your home directory (try ls -a !!!). There we go in simple
   steps:
	
   a) Edit .bbbuttonrc file and type in the bbbuttons you wish 
      to create.
      
      i.e your .bbbuttonrc should contain something like this:
      #--File .bbbuttonrc start---------------------
	#------------------------------------------------#
	#        BBButton Configuration file             #
	# Format: <path to XPM image> <program to run>   #
	#------------------------------------------------#

	[BUTTON_NO]
	MaxButtons   =   20

	[LOAD]
	/root/.DesktopIcons/opera.xpm opera
	/usr/local/shared/icons/cd1.xpm /usr/local/bin/exampleBin
	#/root/.DesktopIcons/bcb.xpm bash --login startbcb   
	/root/.DesktopIcons/gambas.xpm gambas
	/root/.DesktopIcons/xmms.xpm xmms
	/root/.DesktopIcons/mail.xpm sylpheed

      #--File .bbbuttonrc end------------------------

   b) You need to load bbbutton itself through xinitrc.
      
      i.e your xinitrc should contain this:
      --File xinitrc start-
       bbbutton &
       exec blackbox
      --File xinitrc end--- 
      
4. Requirements

    - The image must be of type 'xpm' and size '48x48'
      ( XV, ImageMagic and Gimp are some applications 
        that could do that for you)

5. Licencing

    This program is released under the LGPL license.
    A copy can be found in this distribution in the file 
    COPYING. 

6. Author(s)

    Konstantinos Mantzaris <GODoal@hotmail.com>
    Stathis <p00pie@sourceforge.net>
    
    - Tnx to John Vlahogianis for the 'Stella Artois' supplies
      during the completion of this section. 
    - MBill for the FRAME hints..
    - Extra tnx to Konstantinos Peletidis for porting the 
      Makefile to compile under FreeBSD.
    - Tnx to Kyriakos Oikonomakos for fixing and testing the 
      portability on FreeBSD

7. ENJOY  };)



CHANGES
#### 08/03/2005 ####
- the application commands are now by default appended a "&" 
  at the end, if you include a "&" at the end an error will be
  raised and the application will not run!

- Major code cleanups. The application now runs in one thread.

#### 06/03/2005 ####
- New configuration parser
  
- Maximum limit of buttons to load is no more hardcoded. It
  can be tweaked to your liking through ~/.bbbuttonrc 
  
- Comments on the configuration file and null lines are
  disregarded instead of crashing the application

#### 15/01/2003 ####
- Implemented support for transparent xpm's.
- Tide up rest of the code.

#### 10/01/2003 ####
- Tide up some of the code and changed that README to
  something more user friendly.
   
#### 29/11/2002 ####
- Fixed mouse click problem. Only left mouse button should be
  used to press a button.

#### 09/11/2002 ####
- Modified the Makefile so that the code compiles under FreeBSD.
- Corrected minor typos in the CHANGES file.

#### 29/10/2002 ####
- Changed icon fonts so they adopt the wm's background
  color. Anyway the click efect is much better now (I think).

#### 23/09/2002 ####
- Changed most of the code to object oriented.
- Implemented a configuration file (.bbbuttonrc), so we 
  don't have to deal with xinitrc.

#### 26/07/2002 ####
- Used some sizeof() on the strncpy() for security
  purposes.
- Deleted the huge 128 array length value from 
  strncpy().
- Tide up the rest code.


bbbrc.h
/*
 * BBButton V 8.0  ,bbbrc.h
 *
 * (C) 09-03-2005 Konstantinos Mantzaris <GODoal@hotmail.com>
 *     09-03-2005 Stathis <p00pie@sourceforge.net>
 * 
 * This software is licensed under LGPL
 *
 * Part of this code is striped from wmNetscapeKiller
 * by Pierre-Olivier <pblumo@free.fr>
 * 
 */

#ifndef bbbrc_h
#define bbbrc_h

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <sys/types.h>
#include <unistd.h>

using namespace std;

#include "bbbutton.h"

class BBBrc
{

public :
    BBBrc(vector <BBButton *> *bbbuttons);
    ~BBBrc();

    vector<BBButton *> *aBBButton;

    void createButtons(Display* display);
    bool configValid(void);

    void setButtonNumber(void);
    int getMaxButtons(void) { return maxButtons; }
    int getButtonNumber(void){ return numButtons; }

protected :
    char * rcName;
    int lines;
    int maxButtons;
    int numButtons;
    char picPath[64];
    char picCMD[64];
    vector <string> holdAll;

    void sTrim(string& StringToModify);

};

#endif


main.cc
/*
 * BBButton V 8.0  ,main.cc
 *
 * (C) 09-03-2005 Konstantinos Mantzaris <GODoal@hotmail.com>
 *     09-03-2005 Stathis <p00pie@sourceforge.net>
 *
 * This software is licensed under LGPL
 *
 * Part of this code is striped from wmNetscapeKiller
 * by Pierre-Olivier <pblumo@free.fr>
 *
 */

#include <X11/Xlib.h>
#include <X11/xpm.h>
#include <X11/Xutil.h>
#include <X11/extensions/shape.h>

#include "bbbutton.h"
#include "bbbrc.h"

int main(int argc, char ** argv)
{

    char * pname = argv[0];
    vector <BBButton *> bbbuttons;

    Display         *display;
    Window          root;
    XEvent          event;
    int             screen;

    if( ( display = XOpenDisplay(NULL) ) == NULL )
    {
        cerr<<"Error: XOpenDisplay "<<endl;
        exit(1);
    }

    root = RootWindow(display, screen = DefaultScreen(display));

    BBBrc *rcParser = new BBBrc(&bbbuttons);
    if(argc == 1)
    {
        cout<<"Loading data from .bbbuttonrc ...\n";
	if (rcParser->configValid()) {
	    if ( rcParser->getButtonNumber() > rcParser->getMaxButtons() ) {
    		cout << "Error: The number of buttons to load do not comply with the\n"
            	     << "       maximum button limit specified in ~/.bbbuttonrc"<<endl;
		exit(1);
	    }
	}
	else exit(1);

        rcParser->createButtons(display);
    }
    else
    {
        cout<<"\n+-----------------------------------------------------------------+\n"
	<<"  "<<pname<<" v0.8 ,by Konstantinos Mantzaris (GODoal@hotmail.com)  \n"
	<<"  Available @ http://bbbutton.sourceforge.net/index.html\n"
        <<"\n"
        <<"  Usage : "<<pname<<"\n"
        <<"+-----------------------------------------------------------------+\n";
	exit(1);
    }

    int i;
    delete rcParser;

    while (1)
    {
	XNextEvent(display, &event);
    	for ( i = 0; i < bbbuttons.size(); i++ )
    	{
	    if ( event.xbutton.window == bbbuttons[i]->iconwin ) {
	            bbbuttons[i]->runEvents(&event);
    	    }
	}
    }
    exit(1);
}


bbbutton.h
/*
 * BBButton V 8.0  ,bbbutton.h
 *
 * (C) 09-03-2005 Konstantinos Mantzaris <GODoal@hotmail.com>
 *     09-03-2005 Stathis <p00pie@sourceforge.net>
 *
 * This software is licensed under LGPL
 *
 * Part of this code is striped from wmNetscapeKiller
 * by Pierre-Olivier <pblumo@free.fr>
 *
 */

#ifndef bbbutton_h
#define bbbutton_h

#include <iostream>
#include <fstream>
#include <string.h>
//#include <stdlib.h>
//#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

#include <X11/Xlib.h>
#include <X11/xpm.h>
#include <X11/Xutil.h>
#include <X11/extensions/shape.h>

using namespace std;

class BBButton
{

public :
    BBButton(Display* display, char* Image, char* Command);
    ~BBButton();

    void runEvents(XEvent* e);

    void BPressed(void);
    void BReleased(void);

    Window iconwin;

protected :
    char xpmPath[64];
    char procCMD[64];

private:
    Display	*display;
    Pixmap	XPM;
    GC		gc;

    char ** xpm_data;

};

#endif



bbbutton.cc
/*
 * BBButton V 8.0  ,bbbutton.cc
 *
 * (C) 09-03-2005 Konstantinos Mantzaris <GODoal@hotmail.com>
 *     09-03-2005 Stathis <p00pie@sourceforge.net>
 *
 * This software is licensed under LGPL
 *
 * Part of this code is striped from wmNetscapeKiller
 * by Pierre-Olivier <pblumo@free.fr>
 *
 */

#include "bbbutton.h"

BBButton :: BBButton( Display* display, char* Image, char* Command )
: display(display)
{
    XWMHints    wmHints;
    XSizeHints  sizeHints;
    XClassHint  classHint;
    Pixmap	pixmask;

    int screen DefaultScreen(display);
    long event_mask = ButtonPressMask|ExposureMask|ButtonReleaseMask;
    
    Window root = RootWindow(display, screen);
    long fgcolor = BlackPixel(display, screen);
    long bgcolor = WhitePixel(display, screen);
    this->gc = XDefaultGC(display, screen);

    strncpy(xpmPath, Image, sizeof(xpmPath));
    strncpy(procCMD, Command, sizeof(procCMD));

    sizeHints.x = 0;
    sizeHints.y = 0;
    sizeHints.width = 64;
    sizeHints.height = 64;

    int depth = 1;

    iconwin = XCreateSimpleWindow(display, root, sizeHints.x,
                                  sizeHints.y,
                                  sizeHints.width,
                                  sizeHints.height,
                                  depth, fgcolor, bgcolor);

    Window parentWindow = XCreateSimpleWindow(display, root, sizeHints.x,
                                       sizeHints.y,
                                       sizeHints.width,
                                       sizeHints.height,
                                       depth, fgcolor, bgcolor);

    sizeHints.flags = USSize | USPosition;
    XSetWMNormalHints(display, parentWindow, &sizeHints);

    wmHints.initial_state = WithdrawnState;
    wmHints.icon_window = iconwin;
    wmHints.icon_x = sizeHints.x;
    wmHints.icon_y = sizeHints.y;
    wmHints.window_group = parentWindow;
    wmHints.flags = StateHint|IconWindowHint|IconPositionHint|WindowGroupHint;
    XSetWMHints(display, parentWindow, &wmHints);

    XpmReadFileToPixmap(display, root, Image, &XPM, &pixmask, NULL);
    XShapeCombineMask(display, iconwin, ShapeBounding, 8, 8, pixmask, ShapeSet);

    XSelectInput(display, iconwin, event_mask);
    XMapWindow(display, parentWindow);
}

BBButton::~BBButton()
{
}

void BBButton :: BPressed()
{
    XCopyArea(display, XPM, iconwin, gc,0,0,48,48,9,9);
    system(procCMD);
}

void BBButton :: BReleased()
{
    XCopyArea(display, XPM, iconwin, gc,0,0,48,48,8,8);
}

void BBButton :: runEvents(XEvent *e)
{

    switch(e->type)
    {

    case Expose:
        //cout << "Released Button" << endl;
        BReleased();
        break;

    case ButtonPress:
        switch(e->xbutton.button )
        {
        case 1: //left click
            //cout << "Left Click" << endl;
            BPressed();
            break;
        case 3: //right click
            //cout << "Right Click" << endl;
            break;
        default:
            break;
        }
        break;

    case MotionNotify:
        BReleased();
	break;
    case ButtonRelease:
    	BReleased();
	break;

    }
}




bbbrc.cc
/*
 * BBButton V 8.0  ,bbbrc.cc
 *
 * (C) 09-03-2005 Konstantinos Mantzaris <GODoal@hotmail.com>
 *     09-03-2005 Stathis <p00pie@sourceforge.net>
 *
 * This software is licensed under LGPL
 *
 * Part of this code is striped from wmNetscapeKiller
 * by Pierre-Olivier <pblumo@free.fr>
 *
 */

#include "bbbrc.h"

BBBrc::BBBrc(vector <BBButton *> *bbbuttons)
: aBBButton(bbbuttons)
{
    char *homedir  = getenv("HOME");
    rcName = new char[strlen(homedir) + 32];
    sprintf(rcName, "%s/.bbbuttonrc", homedir);
}

BBBrc::~BBBrc()
{
}

void BBBrc::sTrim(string& StringToModify)
{
    if(StringToModify.empty())
        return;

    int startIndex = StringToModify.find_first_not_of(" ");
    int endIndex = StringToModify.find_last_not_of(" ");
    string tempString = StringToModify;
    StringToModify.erase();

    StringToModify = tempString.substr(startIndex, (endIndex-startIndex+ 1) );
}


bool BBBrc::configValid(void)
{
    char lBuffer[64];
    int count=0;
    char sHold[3][64];
    string tmpStr;

    ifstream fin(rcName, ios::in);
    if(!fin)
    {
        cout<<"Unable to open .bbbuttonrc for reading.\n";
        exit(1);
    }
    while(!fin.eof())
    {
        fin.getline(lBuffer, 64);
        if(fin.peek() == '[')
        {
            fin.ignore(1);
            fin.getline(lBuffer, 64, ']');
            if (strcmp(lBuffer, "BUTTON_NO") == 0)
                strncpy(sHold[0], lBuffer, 64);
            count++;
            if (strcmp(lBuffer, "LOAD") == 0)
                strncpy(sHold[2], lBuffer, 64);
            count++;
        }
        // If the first Header in the INI is [BUTTON_NO] and
        // [LOAD] has not yet been discovered, and we have never been in
        // here before
        if (strcmp(sHold[0], "BUTTON_NO") == 0 && strcmp(sHold[2], "LOAD") != 0 &&
                strcmp(sHold[1], "MaxButtons") !=0 )
        {
            fin.getline(lBuffer, 64, '=');
            fin.ignore(1024, '\040'); // Remove leading spaces
            tmpStr = lBuffer;
            sTrim(tmpStr);           // Remove trailing spaces
            // If the string exists then get the value
            if ((tmpStr.compare(0,11,"\nMaxButtons")) == 0)
            {
                fin.getline(lBuffer, 64, '\n');
                istringstream fullString(lBuffer);
                fullString>>maxButtons;
                strncpy(sHold[1], "MaxButtons", 64);
                count++;
                break;
            }
        }
    }
    fin.close();
    if(count != 3 )
    {
        cout<<"Error: The configuration file is misconfigured"<<endl;
        return false;
    }
    else
    {
	setButtonNumber();	
        return true;
    }
}

void BBBrc::setButtonNumber()
{
    int count = 0, startFlag=0;
    char lBuffer[64];

    ifstream fin(rcName,ios::in);
    if(!fin){
        cout<<"Unable to open .bbbuttonrc for reading.\n";
        exit(1) ;
    }
    while(!fin.eof()) {
        if (startFlag ==0) {
            if(fin.peek() == '[') {
                fin.ignore(1);
                fin.getline(lBuffer, 64, ']');
                if (strcmp(lBuffer, "LOAD") == 0)
                    startFlag = 1;
            }
            else fin.ignore(1024, '\n');
        }
        else {
            if( fin.peek() == '#' || fin.peek() == '\0' || fin.peek() == '\n') {
                fin.ignore(1024, '\n');
            }
            else {
                if(!fin.eof()) {
                    count++;
                    fin.getline(picPath,64,'\040');
                    fin.getline(picCMD,64,'\n');
		    string sAmper(picCMD);
		    sAmper.append(" &");
		    holdAll.push_back(picPath);
                    holdAll.push_back(sAmper);
                }
            }

        }
    }
    fin.close();
    numButtons = count;
}

void BBBrc::createButtons(Display* display)
{    
	for( int i = 0; i < holdAll.size() - 1; i+=2 ) {
		aBBButton->push_back(new BBButton(display, 
					(char *) holdAll[i].c_str(), 
					(char *) holdAll[i+1].c_str()));
	}
}
